src/shared/discover-image.c

```diff
@@ -1590,6 +1590,21 @@ int image_set_limit(Image *i, uint64_t referenced_max) {
         return 0;
 }
 
+int image_set_pool_limit(uint64_t referenced_max) {
+        const char *dir = "/var/lib/machines";
+        int r;
+
+        r = btrfs_qgroup_set_limit(dir, /* qgroupid = */ 0, referenced_max);
+        if (r < 0)
+                log_debug_errno(r, "Failed to set limit on btrfs quota group for '%s', ignoring: %m", dir);
+
+        r = btrfs_subvol_set_subtree_quota_limit(dir, /* subvol_id = */ 0, referenced_max);
+        if (r < 0)
+                return log_debug_errno(r, "Failed to set subtree quota limit for '%s': %m", dir);
```

Comment:
Please check for btrfs first. Otherwise it gets potentially noisy for no reason and performs useless operations.

Reply:
(btw, ideally we'd pin the dir first, then operate on a single fd, i.e. check its btrfs, apply the two quotas. then close it. instead of going via patch each time. that's a bigger refactoring though).

Reply:
your first suggestion sounds good btw.